Library/Homebrew/extend/os/mac/keg_relocate.rb

```diff
@@ -56,11 +54,21 @@ def fix_dynamic_linkage
           change_install_name(bad_name, new_name, file) unless new_name == bad_name
         end
 
+        # Count duplicate rpaths. We need to keep track of this ourselves
+        # because the MachO data is cached and this cache is not updated
+        # after modification with #delete_rpath.
```

Comment:
This feels like something it'd be good to update ruby-macho to do (or provide an API to do). CC @woodruffw for thoughts.

Reply:
I _think_ the caching happens on our side, here:

https://github.com/Homebrew/brew/blob/96cf564ca9355008897cac88ae1bb584098922f6/Library/Homebrew/os/mac/mach.rb#L14-L17

Reply:
Gotcha. In that case: perhaps just bust the cache after `delete_rpath`?

Reply:
Hmm. Not sure of a good way to. `MachO.open` can get pretty expensive, so we don't want to have to do that again. We could probably keep our own data about `rpaths` here instead of delegating to `MachO::MachOFile` (or whatever `MachO.open` returns). But that'll be a bunch of extra code which I'm not sure is needed right now.

Reply:
This is probably a case of confusing mismatched state -- `Keg#delete_rpath` used `MachO::Tools::delete_rpath`, and not `MachOFile#delete_rpath`. The former is a one-shot API that doesn't used our cached `@macho`, since it takes (and writes) the FS path directly.

`MachOFile#delete_rpath` *would* correctly update the cache, since it performs an internal update unless `repopulate: false` is passed.

Reply:
I think `delete_rpath` did use to be delegated to `MachOFile` previously, except it didn't work. I think that was because it was missing `#write!` call, though.

Reply:
Yep, that's probably why -- we *do* update the in-memory `@macho` implicitly but never the on-disk copy until the user explicitly asks for it. I can look into refactoring `delete_rpath` to use `MachOFile` this evening (EST).

Reply:
I've actually already done some of the refactoring. But it's mostly superficial: I just replaced `MachO::Tools.delete_rpath` with `MachOFile#delete_rpath` + `MachOFile#write!`.

There's probably a more comprehensive refactoring which delays the `#write!` call, and does the same with the other uses of `MachO::Tools`.

Reply:
#12750

But feel free to open your own PR if you figure out a nicer way to do this.

Reply:
That SGTM to me for now! You're probably right that we can defer the `#write!`, but that'll probably take more refactoring consideration. I'll review your changes now.